src/bindings/python/flux/job/executor.py

```diff
@@ -0,0 +1,524 @@
+            time.sleep(self.__poll_interval)
+        while self.__jobspecs_to_submit:
+            try:
+                jobspec, user_future = self.__jobspecs_to_submit.popleft()
+            except IndexError:
+                continue
+            if user_future.set_running_or_notify_cancel():
+                try:
+                    submit_async(self._flux_handle, jobspec, waitable=True).then(
+                        self._submission_callback, user_future
+                    )
+                except OSError as os_error:
+                    user_future.set_exception(os_error)
+                else:
+                    self._remaining_flux_futures += 1
+
+    def _submission_callback(self, submission_future, user_future):
+        """Callback invoked when a jobid is ready for a submitted jobspec.
+
+        Overridden by subclasses.
+        """
+        jobid = submit_get_id(submission_future)
+        user_future.set_jobid(jobid)
+        wait_async(self._flux_handle, jobid).then(
+            self.__complete_user_future, user_future
+        )
+
+    def __complete_user_future(self, wait_future, user_future):
+        """Callback invoked when a job has completed."""
+        result = wait_future.get_status()
+        self._remaining_flux_futures -= 1
+        if result.success:
+            user_future.set_result(None)
+        else:
+            user_future.set_exception(
+                JobFailure(f"Job exited abnormally: {result.errstr}")
+            )
+
+
+class FluxEventExecutorThread(FluxExecutorThread):
+    """Subclass of FluxExecutorThread that updates events on futures."""
+
+    def _submission_callback(self, submission_future, user_future):
+        """Callback invoked when a jobid is ready for a submitted jobspec."""
+        jobid = submit_get_id(submission_future)
+        user_future.set_jobid(jobid)
+        event_watch_async(self._flux_handle, jobid).then(
+            self.__update_state, user_future
+        )
+
+    def __update_state(self, event_future, user_future):
+        event = event_future.get_event()
+        if event is not None:
+            if event.name in user_future.EVENTS:
+                user_future.set_event(event)
+        else:  # no more events
+            self._remaining_flux_futures -= 1
+
+
+class FluxExecutorFuture(concurrent.futures.Future):
+    """A concurrent.futures.Future subclass that represents a single Flux job.
+
+    Provides the ``jobid`` and ``add_jobid_callback`` methods for retrieving the
+    Flux jobid of the underlying job.
+    """
+
+    def __init__(self, owning_thread_id, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.__jobid_condition = threading.Condition()
+        self.__jobid = None
+        self.__jobid_callbacks = []
+        # Thread.ident of thread tasked with completing this future
+        self.__owning_thread_id = owning_thread_id
+
+    def set_jobid(self, jobid):
```

Comment:
I'm happy either way. I'm just following the `concurrent.futures` example, which makes the methods public but documents them as "only intended to be used by Executor implementations and unit tests." But if you don't expect users to make their own executors that use these futures (or tests around them) then yeah we could make them private.

Reply:
Ah. I wonder why they (concurrent.futures) did that. I think my preference here is to prefix them, to avoid any confusion for users. We can always remove the prefix in the future if we find a compelling reason.  Always easier to take something private and make it public than vice-a-versa.

Reply:
Done. Yeah I bet `concurrent.futures` did it because technically the methods are part of the public API, they just have very specific uses. But as you mentioned, when users want to subclass the executors or design their own, it would be very easy to make the methods public again.